pulser-simulation/pulser_simulation/qutip_backend.py

```diff
@@ -88,3 +96,138 @@ def run(
+                self._sim_obj.total_duration_ns
+            ),
+        )
+        if self._config.initial_state:
+            self._sim_obj.set_initial_state(
+                self._config.initial_state.to_qobj()
+            )
+
+    def run(self) -> Results:
+        """Executes the sequence on the backend."""
+        res = Results(
+            atom_order=tuple(self._sequence.qubit_info),
+            total_duration=self._sim_obj.total_duration_ns,
+        )
+        eigenstates = self._sim_obj.samples_obj.eigenbasis
+
+        if (
+            "amplitude" not in self._sim_obj.config.noise
+            or self._sim_obj.config.amp_sigma == 0.0
+        ) and (
+            "SPAM" not in self._sim_obj.config.noise
+            or self._sim_obj.config.eta == 0
+        ):
+            # A single run is needed, regardless of self.config.runs
+            _sim_res = self._sim_obj._run_solver(progress_bar=False)
+            assert isinstance(_sim_res, CoherentResults)
+
+            for qutip_res in _sim_res:
+                t_ = qutip_res.evaluation_time
+                state = QutipState(qutip_res.state, eigenstates=eigenstates)
+                ham: QutipOperator = QutipOperator(
+                    self._sim_obj.get_hamiltonian(t_ * res.total_duration),
+                    eigenstates=eigenstates,
+                )
+                for obs in self._config.observables:
+                    obs(
+                        config=self._config,
+                        t=t_,
+                        state=state,
+                        hamiltonian=ham,
+                        result=res,
+                    )
+
+        else:
+            density_matrices: dict[float, qutip.Qobj] = {}
+            total_reps = 0
+            for cleanres_noisyseq, reps in self._sim_obj._noisy_runs(
+                progress_bar=False
+            ):
+                total_reps += reps
+                for index, qutip_res in enumerate(cleanres_noisyseq):
+                    t_ = qutip_res.evaluation_time
+
+                    density_matrices.setdefault(
+                        t_,
+                        qutip.tensor(
+                            [
+                                qutip.Qobj(np.zeros((2, 2)))
+                                for _ in range(
+                                    self._sim_obj._hamiltonian._size
+                                )
+                            ]
+                        ),
+                    )
+
+                    if qutip_res.state.isoper:
+                        density_matrices[t_] += reps * qutip_res.state
+                    else:
+                        density_matrices[t_] += (
+                            reps * qutip_res.state * qutip_res.state.dag()
+                        )
+
+            density_matrices = {
+                t: m / total_reps for t, m in density_matrices.items()
+            }
+
+            for t, dm in density_matrices.items():
```

Comment:
I might be misunderstanding the code, but should you aggregrate the `density_matrices` dict as
```
sampled_state = qutip.tensor(
                            [
                                qutip.Qobj(np.zeros((2, 2)))
                                for _ in range(
                                    self._sim_obj._hamiltonian._size
                                )
                            ]
                        )
for weight, dm in density_matrices.items():
    sampled_state += weight * dm
```
and then call observables on that?

Reply:
every shot gives you a density matrix for all times, so there are as many accumulating density matrices as there are sampling times and every new noisy run means adding new DMs for each time.
It would aggregate as you describe (for a given time) if I were to store every DM for every time and every run, and postprocess, but that would use too much memory, I think, without providing additional value afaics

Reply:
ah, I see, you do this aggregation in the previous loop, on the results yielded from `noisy_runs`. Never mind then.